server/src/main/java/org/elasticsearch/action/admin/indices/diskusage/IndexDiskUsageAnalyzer.java

```diff
@@ -0,0 +1,665 @@
+                    break;
+                case BINARY:
+                    final BinaryDocValues binary = docValuesReader.getBinary(field);
+                    while (binary.nextDoc() != DocIdSetIterator.NO_MORE_DOCS) {
+                        cancellationChecker.logEvent();
+                        binary.binaryValue();
+                    }
+                    break;
+                case SORTED:
+                    final SortedDocValues sorted = docValuesReader.getSorted(field);
+                    while (sorted.nextDoc() != DocIdSetIterator.NO_MORE_DOCS) {
+                        cancellationChecker.logEvent();
+                        sorted.ordValue();
+                    }
+                    sorted.lookupOrd(0);
+                    sorted.lookupOrd(sorted.getValueCount() - 1);
+                    break;
+                case SORTED_SET:
+                    final SortedSetDocValues sortedSet = docValuesReader.getSortedSet(field);
+                    while (sortedSet.nextDoc() != DocIdSetIterator.NO_MORE_DOCS) {
+                        while (sortedSet.nextOrd() != SortedSetDocValues.NO_MORE_ORDS) {
+                            cancellationChecker.logEvent();
+                        }
+                    }
+                    sortedSet.lookupOrd(0);
+                    sortedSet.lookupOrd(sortedSet.getValueCount() - 1);
+                    break;
+                default:
+                    assert false : "Unknown docValues type [" + dvType + "]";
+                    throw new IllegalStateException("Unknown docValues type [" + dvType + "]");
+            }
+            stats.addDocValues(field.name, directory.getBytesRead());
+        }
+    }
+
+    void analyzePostings(SegmentReader reader, IndexDiskUsageStats stats) throws IOException {
+        // TODO: FieldsReader has stats() which might contain the disk usage infos
+        // Also, can we track the byte reader per field extension to avoid visiting terms multiple times?
+        FieldsProducer postingsReader = reader.getPostingsReader();
+        if (postingsReader == null) {
+            return;
+        }
+        postingsReader = postingsReader.getMergeInstance();
+        PostingsEnum postings = null;
+        for (FieldInfo field : reader.getFieldInfos()) {
+            if (field.getIndexOptions() == IndexOptions.NONE) {
+                continue;
+            }
+            cancellationChecker.checkForCancellation();
+            directory.resetBytesRead();
+            final Terms terms = postingsReader.terms(field.name);
+            if (terms == null) {
+                continue;
+            }
+
+            // It's expensive to look up every term and visit every document of the postings lists of all terms.
+            // As we track the min/max positions of read bytes, we just visit the two ends of a partition containing
+            // the data. We might some small parts of the data, but it's an good trade-off to speed up the process.
```

Comment:
```suggestion
            // the data. We might miss some small parts of the data, but it's an good trade-off to speed up the process.
```

Reply:
Given this comment, I was expected the code to only look up `terms.getMin()` and `terms.getMax()`, do we actually need to iterate over the entire terms dict?